arch/riscv/core/pmp.c

```diff
@@ -479,11 +530,35 @@ void z_riscv_pmp_init(void)
 		      Z_RISCV_STACK_GUARD_SIZE,
 		      pmp_addr, pmp_cfg, ARRAY_SIZE(pmp_addr));
 
+#endif /* CONFIG_MULTITHREADING */
+#endif
+
+#ifdef CONFIG_MEM_ATTR
+	/*
+	 * Set the memory attribute region as temporary PMP entries for early
+	 * kernel initialization. This provides essential protection before
+	 * the kernel mode memory attribute permission is fully operational.
+	 */
+	unsigned int attr_cnt = set_pmp_mem_attr(&index, pmp_addr, pmp_cfg, ARRAY_SIZE(pmp_addr));
+
+	/*
+	 * This early, we want to protect unlock PMP entries as soon as
+	 * possible. But we need a temporary default "catch all" PMP entry for
+	 * MPRV to work. Later on, this entry will be set for each thread by
+	 * z_riscv_pmp_kernelmode_prepare().
+	 */
+	set_pmp_mprv_catchall(&index, pmp_addr, pmp_cfg, ARRAY_SIZE(pmp_addr));
+
 	/* Write those entries to PMP regs. */
 	write_pmp_entries(0, index, true, pmp_addr, pmp_cfg, ARRAY_SIZE(pmp_addr));
-#endif /* CONFIG_MULTITHREADING */
+
+	/* Activate our non-locked PMP entries for m-mode */
+	csr_set(mstatus, MSTATUS_MPRV);
+
+	/* And forget about that tempory entries as we won't need it later */
+	index -= attr_cnt + 1;
```

Comment:
Since there is no call to `z_riscv_pmp_kernelmode_prepare()` and `z_riscv_pmp_kernelmode_enable()` when CONFIG_MULTITHREADING=n, PMP MEM_ATTR should be global entry in this case.


```
#ifdef CONFIG_MULTITHREADING
	/* And forget about that tempory entries as we won't need it later */
	index -= attr_cnt + 1;
#else
	ARG_UNUSED(attr_cnt);
#endif /* CONFIG_MULTITHREADING */
```